Add --username option to Clip command.

sjamesr opened the pull request:
Adds a `--username` option to the clip command, so that usernames may be copied to the clipboard in the same way as passwords.

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ New feature (non-breaking change which adds functionality)

## Testing strategy
Tested manually, added unit test.

## Checklist:
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**

louib commented on src/cli/Clip.cpp:
```diff
@@ -32,11 +32,16 @@ const QCommandLineOption Clip::TotpOption = QCommandLineOption(QStringList() <<
 {
     name = QString("clip");
-    description = QObject::tr("Copy an entry's password to the clipboard.");
+    description = QObject::tr("Copy an entry's password or username to the clipboard.");
```
@sjamesr this description is already outdated, since we allow the current totp to be copied. Maybe we should update it to `Copy an entry's field`?

sjamesr commented:
Hey @louib, thanks for taking a look! I updated the clip command to take an arbitrary attribute name (it defaults to "password"). I removed the `--totp` flag, but I could add that back as an alias for `-a totp`.

louib commented:
@sjamesr I think it's a good idea to keep the `--totp` as an alias!

sjamesr commented:
Ok, `--totp` is now retained as an alias and the documentation is updated a bit.

louib commented:
@sjamesr could you also add the new option to the manpage please?

sjamesr commented:
Done, please take a look

On Fri, Nov 29, 2019 at 10:03 AM louib <notifications@github.com> wrote:

> @sjamesr <https://github.com/sjamesr> could you also add the new option
> to the manpage please?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/keepassxreboot/keepassxc/pull/3947?email_source=notifications&email_token=ABDNGA57VMZETQB2RCINXALQWFKOZA5CNFSM4JSXM232YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEFPLZ2Y#issuecomment-559856875>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABDNGA6OHLQBNIQBC4YLF3DQWFKOZANCNFSM4JSXM23Q>
> .
>

louib commented on src/cli/Clip.cpp:
```diff
@@ -70,28 +77,39 @@ int Clip::executeWithDatabase(QSharedPointer<Database> database, QSharedPointer<
         return EXIT_FAILURE;
     }
 
+    QString selectedAttribute = parser->value(AttributeOption);
```
@sjamesr last nitpick:
Should we raise if both `-t` and `-a` are used?
e.g. `clip -t -a some_other_field`

louib replied:
Also, will `selectedAttribute` be populated if the `-t` option is used?

sjamesr replied:
Added an error message if more than one attribute is requested, or if `--totp` and `-a` are both specified.

Well spotted about `selectedAttribute` not being populated, I added a test for that too.

louib commented on src/cli/Clip.cpp:
```diff
@@ -70,28 +77,39 @@ int Clip::executeWithDatabase(QSharedPointer<Database> database, QSharedPointer<
-        value = entry->password();
+        for (const QString& key : entry->attributes()->keys()) {
+            if (key.compare(selectedAttribute, Qt::CaseSensitivity::CaseInsensitive) == 0) {
+                value = entry->attribute(key);
```
@sjamesr why not use `value = entry->attributes()->value(selectedAttribute)`?

sjamesr replied:
I wanted case insensitivity, what do you think?

louib replied:
I just realized that means the attribute name is case sensitive when we use the `-a` option in the `show` CLI command :grimacing: 

Maybe we could make it case insensitive there too and extract that logic in a function. Not sure `EntryAttribute::` is the right place though, maybe in `cli/Utils`?

sjamesr replied:
I extracted this code to Utils::findAttributes and updated `clip` and `show` to use it. If the user-specified attribute name is ambiguous, an error message is displayed by both `clip` and `show`.

droidmonkey commented:
@louib is this one ready to go?

louib commented:
@sjamesr can you do another rebase please? @droidmonkey yeah this one is ready to go after the rebase.

sjamesr commented:
Should be good to go now, thanks.

On Mon, Jan 27, 2020 at 6:46 AM louib <notifications@github.com> wrote:

> @sjamesr <https://github.com/sjamesr> can you do another rebase please?
> @droidmonkey <https://github.com/droidmonkey> yeah this one is ready to
> go after the rebase.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/keepassxreboot/keepassxc/pull/3947?email_source=notifications&email_token=ABDNGA7YFTELTDJDXCKX6PLQ73XVHA5CNFSM4JSXM232YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJ7YL2Y#issuecomment-578782699>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABDNGA73TAZKDGQCCZ6SQO3Q73XVHANCNFSM4JSXM23Q>
> .
>

louib commented on src/cli/Show.cpp:
```diff
@@ -79,31 +81,39 @@ int Show::executeWithDatabase(QSharedPointer<Database> database, QSharedPointer<
     // If no attributes specified, output the default attribute set.
     bool showDefaultAttributes = attributes.isEmpty() && !showTotp;
-    if (attributes.isEmpty() && !showTotp) {
+    if (showDefaultAttributes) {
```
@sjamesr :+1:

sjamesr commented:
@louib no problem :)

I added a test that exercises the current behavior.

louib commented on tests/TestCli.cpp:
```diff
@@ -538,6 +543,22 @@ void TestCli::testClip()
+        QByteArray("ERROR: attribute TESTAttribute1 is ambiguous, it matches TestAttribute1 and testattribute1.\n"));
+
+    // Test handling of both TOTP and --attribute. Currently, this will cause
+    // --attribute to be ignored.
```
@sjamesr Isn't it better to raise in that case? I feel like that should be an invalid invocation

sjamesr replied:
Done, please take a look.
